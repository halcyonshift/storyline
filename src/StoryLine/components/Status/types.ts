import {
    CharacterModel,
    ItemModel,
    LocationModel,
    NoteModel,
    SectionModel,
    WorkModel
} from '@sl/db/models'

export type StatusProps = {
    model: CharacterModel | ItemModel | LocationModel | NoteModel | SectionModel | WorkModel
}
