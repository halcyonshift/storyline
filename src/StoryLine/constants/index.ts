export const YYYYMMDD = 'yyyy-MM-dd'
