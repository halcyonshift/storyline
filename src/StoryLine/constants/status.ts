import { useTranslation } from 'react-i18next'
import { colors } from '@sl/theme/utils'

export type StatusType = 'TODO' | 'DRAFT' | 'REVIEW' | 'COMPLETE' | 'ARCHIVE'

export const Status = Object.freeze({
    TODO: 'TODO',
    DRAFT: 'DRAFT',
    REVIEW: 'REVIEW',
    COMPLETE: 'COMPLETE',
    ARCHIVE: 'ARCHIVE'
})

export const status = (status?: StatusType) => {
    const { t } = useTranslation()

    const statusList = {
        [Status.TODO]: {
            label: t('status.todo'),
            color: colors.red
        },
        [Status.DRAFT]: {
            label: t('status.draft'),
            color: colors.blue
        },
        [Status.REVIEW]: {
            label: t('status.review'),
            color: colors.yellow
        },
        [Status.COMPLETE]: {
            label: t('status.complete'),
            color: colors.green
        },
        [Status.ARCHIVE]: {
            label: t('status.archive'),
            color: colors.slate
        }
    }

    return status ? statusList[status] : status
}
