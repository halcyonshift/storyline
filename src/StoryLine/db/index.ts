/** @format */

import { Database } from '@nozbe/watermelondb'
import LokiJSAdapter from '@nozbe/watermelondb/adapters/lokijs'

import schema from './schema'
// import migrations from './migrations'

import {
    AnnotationModel,
    CharacterModel,
    ItemModel,
    LocationModel,
    NoteModel,
    ProjectModel,
    SectionModel,
    StatisticModel
} from './models'

const adapter = new LokiJSAdapter({
    schema,
    // migrations,
    useWebWorker: false,
    useIncrementalIndexedDB: true,
    dbName: 'storyline',
    onQuotaExceededError: (error) => {
        console.log(error)
        // ToDo - Error catch
        // Browser ran out of disk space -- offer the user to reload the app or log out
    },
    onSetUpError: (error) => {
        console.log(error)
        // ToDo - Error catch
        // Database failed to load -- offer the user to reload the app or log out
    },
    extraIncrementalIDBOptions: {
        onDidOverwrite: () => {
            // Called when this adapter is forced to overwrite contents of IndexedDB.
            // This happens if there's another open tab of the same app that's making changes.
            // Try to synchronize the app now, and if user is offline, alert them that if they
            // close this tab, some data may be lost
        },
        onversionchange: () => {
            // database was deleted in another browser tab (user logged out), so we must make sure
            // we delete it in this tab as well - usually best to just refresh the page
            // if (checkIfUserIsLoggedIn()) {
            //  window.location.reload()
            //}
        }
    }
})

// Then, make a Watermelon database from it!
const database = new Database({
    adapter,
    modelClasses: [
        AnnotationModel,
        CharacterModel,
        ItemModel,
        LocationModel,
        NoteModel,
        ProjectModel,
        SectionModel,
        StatisticModel
    ]
})

export default database
