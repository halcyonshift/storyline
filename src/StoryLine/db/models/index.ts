export { default as AnnotationModel } from './AnnotationModel'
export { default as CharacterModel } from './CharacterModel'
export { default as ItemModel } from './ItemModel'
export { default as LocationModel } from './LocationModel'
export { default as NoteModel } from './NoteModel'
export { default as WorkModel } from './WorkModel'
export { default as SectionModel } from './SectionModel'
export { default as StatisticModel } from './StatisticModel'
