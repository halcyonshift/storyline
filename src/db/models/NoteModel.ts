import { Model, Relation } from '@nozbe/watermelondb'
import { Associations } from '@nozbe/watermelondb/Model'
import { date, field, readonly, relation, text } from '@nozbe/watermelondb/decorators'

import CharacterModel from './CharacterModel'
import ItemModel from './ItemModel'
import LocationModel from './LocationModel'


export default class NoteModel extends Model {
  static table = 'note'
  public static associations: Associations = {
    character: { type: 'belongs_to', key: 'character_id' },
    item: { type: 'belongs_to', key: 'item_id' },
    location: { type: 'belongs_to', key: 'location_id' }
  }

  @text('title') name!: string
  @text('body') body!: string
  @date('date') date!: string
  @text('url') url!: string
  @field('image') image!: string
  @readonly @date('created_at') createdAt!: Date
  @readonly @date('updated_at') updatedAt!: Date

  @relation('character', 'character_id') project!: Relation<CharacterModel>
  @relation('item', 'item_id') item!: Relation<ItemModel>
  @relation('location', 'location_id') location!: Relation<LocationModel>
}