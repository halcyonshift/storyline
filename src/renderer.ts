import { init } from '@sentry/electron/renderer'
import { init as reactInit } from '@sentry/react'

if (parseInt(process.env.DEBUG) === 0) {
    init({ dsn: process.env.SENTRY }, reactInit)
}

import '@fontsource/roboto/300.css'
import '@fontsource/roboto/400.css'
import '@fontsource/roboto/500.css'
import '@fontsource/roboto/700.css'
import './index.css'
import './StoryLine'
