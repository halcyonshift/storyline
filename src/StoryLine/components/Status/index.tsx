import Button from '@mui/material/Button'
import ButtonGroup from '@mui/material/ButtonGroup'

import { Status as Options, type StatusType } from '@sl/constants/status'
import { useTranslation } from 'react-i18next'
import { status } from '@sl/theme/utils'
import { StatusProps } from './types'

const Status = ({ model }: StatusProps) => {
    const { t } = useTranslation()

    return (
        <ButtonGroup size='small' variant='outlined'>
            {Object.keys(Options).map((option: StatusType) => (
                <Button
                    sx={{
                        backgroundColor: option === model.status ? status(option, 400).color : ''
                    }}
                    color='inherit'
                    key={option}
                    value={option}
                    onClick={() => model.updateStatus(option)}>
                    {t(`constant.status.${option}`)}
                </Button>
            ))}
        </ButtonGroup>
    )
}

export default Status
