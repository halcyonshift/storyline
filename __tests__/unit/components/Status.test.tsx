import userEvent from '@testing-library/user-event'
import renderer from 'react-test-renderer'
import { Status as Options } from '@sl/constants/status'
import Status from '@sl/components/Status'
import database from '@sl/db'
import { ItemModel, WorkModel } from '@sl/db/models'
import { render, screen } from '../test-utils'

describe('<Status />', () => {
    let work: WorkModel
    let item: ItemModel

    beforeAll(async () => {
        await database.write(async () => {
            await database.unsafeResetDatabase()
        })

        work = await database.write(async () => {
            return database.get<WorkModel>('work').create((work) => {
                work.title = 'Test'
            })
        })

        item = await work.addItem({ name: 'Test Item' })

        render(<Status model={item} />)
    })

    it('it updates the model status on click', async () => {
        const buttons = screen.getAllByRole('button')

        await userEvent.click(buttons[0])
        expect(item.status).toBe(Options.TODO)

        await userEvent.click(buttons[1])
        expect(item.status).toBe(Options.DRAFT)

        await userEvent.click(buttons[2])
        expect(item.status).toBe(Options.REVIEW)

        await userEvent.click(buttons[3])
        expect(item.status).toBe(Options.COMPLETE)

        await userEvent.click(buttons[4])
        expect(item.status).toBe(Options.ARCHIVE)
    })

    it('renders correctly', () => {
        const tree = renderer.create(<Status model={item} />).toJSON()
        expect(tree).toMatchSnapshot()
    })
})
