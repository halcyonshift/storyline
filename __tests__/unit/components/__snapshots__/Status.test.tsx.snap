// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Status /> renders correctly 1`] = `
<div
  className="MuiButtonGroup-root MuiButtonGroup-outlined css-iajp3t-MuiButtonGroup-root"
  role="group"
>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedInherit MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-colorInherit MuiButton-disableElevation MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary MuiButton-root MuiButton-contained MuiButton-containedInherit MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-colorInherit MuiButton-disableElevation MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary css-q1c37s-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
    value="TODO"
  >
    constant.status.TODO
  </button>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedInherit MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-colorInherit MuiButton-disableElevation MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary MuiButton-root MuiButton-contained MuiButton-containedInherit MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-colorInherit MuiButton-disableElevation MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary css-q1c37s-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
    value="DRAFT"
  >
    constant.status.DRAFT
  </button>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedInherit MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-colorInherit MuiButton-disableElevation MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary MuiButton-root MuiButton-contained MuiButton-containedInherit MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-colorInherit MuiButton-disableElevation MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary css-q1c37s-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
    value="REVIEW"
  >
    constant.status.REVIEW
  </button>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedInherit MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-colorInherit MuiButton-disableElevation MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary MuiButton-root MuiButton-contained MuiButton-containedInherit MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-colorInherit MuiButton-disableElevation MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary css-q1c37s-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
    value="COMPLETE"
  >
    constant.status.COMPLETE
  </button>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-disableElevation MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-disableElevation MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary css-zerb3r-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
    value="ARCHIVE"
  >
    constant.status.ARCHIVE
  </button>
</div>
`;
